contracts/FungibleToken.cdc

```diff
@@ -102,136 +112,116 @@ pub contract interface FungibleToken {
                 return {}
             }
         }
-    }
-
-    /// The interface that contains the `balance` field of the Vault
-    /// and enforces that when new Vaults are created, the balance
-    /// is initialized correctly.
-    ///
-    pub resource interface Balance {
-
-        /// The total balance of a vault
-        ///
-        pub var balance: UFix64
-
-        init(balance: UFix64) {
-            post {
-                self.balance == balance:
-                    "Balance must be initialized to the initial balance"
-            }
-        }
-
-        /// Function that returns all the Metadata Views implemented by a Fungible Token
-        ///
-        /// @return An array of Types defining the implemented views. This value will be used by
-        ///         developers to know which parameter to pass to the resolveView() method.
-        ///
-        pub fun getViews(): [Type] {
-            return []
-        }
 
-        /// Function that resolves a metadata view for this fungible token by type.
-        ///
-        /// @param view: The Type of the desired view.
-        /// @return A structure representing the requested view.
-        ///
-        pub fun resolveView(_ view: Type): AnyStruct? {
-            return nil
+        /// Checks if the given type is supported by this Vault
+        access(all) view fun isSupportedVaultType(type: Type): Bool {
+            return self.getSupportedVaultTypes()[type] ?? false
         }
-    }
-
-    /// The resource that contains the functions to send and receive tokens.
-    /// The declaration of a concrete type in a contract interface means that
-    /// every Fungible Token contract that implements the FungibleToken interface
-    /// must define a concrete `Vault` resource that conforms to the `Provider`, `Receiver`,
-    /// and `Balance` interfaces, and declares their required fields and functions
-    ///
-    pub resource Vault: Provider, Receiver, Balance {
-
-        /// The total balance of the vault
-        pub var balance: UFix64
 
-        // The conforming type must declare an initializer
-        // that allows providing the initial balance of the Vault
-        //
-        init(balance: UFix64)
-
-        /// Subtracts `amount` from the Vault's balance
+        /// withdraw subtracts `amount` from the Vault's balance
         /// and returns a new Vault with the subtracted balance
         ///
-        /// @param amount: The amount of tokens to be withdrawn from the vault
-        /// @return The Vault resource containing the withdrawn funds
-        ///
-        pub fun withdraw(amount: UFix64): @Vault {
+        access(Withdraw) fun withdraw(amount: UFix64): @{Vault} {
             pre {
                 self.balance >= amount:
                     "Amount withdrawn must be less than or equal than the balance of the Vault"
             }
             post {
+                result.getType() == self.getType(): "Must return the same vault type as self"
```

Comment:
In same vein as comment below, I'm not entirely sure this should be included as a check at the interface level. On one hand, I can understand wanting to support consistent caller expectation. On the other, I'm wondering if it's a remnant of a single-type vault design, conflicting with the idea that multiple vault types are supported

Reply:
like my other comment, I think you're misunderstanding what it means to support multiple vault types. Vaults HAVE to return a vault of their type from a withdraw. If we didn't include this, people could trick callers by returning a different vault type and the caller wouldn't know unless they explicitly checked, which is way too risky. If someone wants withdraw to be able to return multiple types, they can just define their own different resource the implements `FungibleToken.Provider`

Reply:
I see, that makes sense. Thanks for clarifying!